doc/developer/design/20231110_aws_connections.md

```diff
@@ -0,0 +1,129 @@
+User then creates a connection in Materialize with that arn.
+```
+CREATE CONNECTION aws_conn TO AWS (
+  ASSUME ROLE ARN = 'arn:aws:iam::001234567890:role/MaterializeConn'
+);
+```
+On Materialize's end we generate an External ID and create an entry in a new `mz_aws_connections` table.
+Users should be able to get the external_id and the principal by querying this table.
+```
+SELECT principal, external_id
+FROM mz_aws_connections aws
+JOIN mz_connections c ON aws.id = c.id
+WHERE c.name = 'aws_conn';
+```
+
+Users should then be able to add a trust policy in their AWS account to give access to Materialize.
+```
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "",
+      "Effect": "Allow",
+      "Principal": {
+        "AWS": "<principal>"
+      },
+      "Action": "sts:AssumeRole",
+      "Condition": {
+        "StringEquals": {
+          "sts:ExternalId": "<external_id>"
+        }
+      }
+    }
+  ]
+}
+```
+
+## Implementation
+We already have some [existing code](https://github.com/MaterializeInc/materialize/blob/v0.77.1/src/storage-types/src/connections/aws.rs), which can be re-used to
+handle the AssumeRole scenario as well.
+
+In the parser we already support `ROLE ARN` option which can be renamed to `ASSUME ROLE ARN` to better reflect the usage.
+We will need to add an `ASSUME ROLE SESSION NAME` option as well.
+
+The AWS Connection create statement with all the options will now look like:
+```sql
+CREATE CONNECTION <connector_name> TO AWS (
+  ACCESS KEY ID [[=] <value>], -- existing
+  SECRET ACCESS KEY [[=] SECRET <value>], -- existing
+  TOKEN [[=] SECRET <value>], -- existing
+  ASSUME ROLE ARN [[=] <value>], -- rename existing from ROLE ARN
+  ASSUME ROLE SESSION NAME [[=] <value>], -- new option
+  ENDPOINT [[=] <value>], -- existing
+  REGION [[=] <value>] -- existing
+);
+```
+
+#### External ID and principal
+We already have external ID prefix and principal provided in the catalog state.
+The connection ID will be appended to the external ID prefix to get the complete
+External ID and then stored in the new `mz_aws_connections` table for that connection ID.
+
+## Rollout and Testing
+We should put the AWS connection behind a feature flag.
+
+#### Testing During development
+Write cloudtests to test out the different AWS Connections. For the AssumeRole AWS Connection we should test that we are able to get temporary credentials.
+
+#### Testing after code is merged
+Switch on the feature flag for the staging environment and create an AWS Connection in staging followed by `VALIDATE CONNECTION`.
+
+## Open questions
+
+#### What should be the behaviour for `VALIDATE CONNECTION` of an AWS connection?
+A `VALIDATE CONNECTION` of an AWS connection with AssumeRole can check if we are
+able to get temporary credentials with the given arn.
+
+In case of tokens though, what should we check in validate? Maybe that's a noop and
+it's fine, because usually tokens will not be used for actual production use cases.
```

Comment:
I am leaning towards
* With AssumeRole, check if we can get temporary credentials
* With tokens we do a noop (unless there's an existing API to validate tokens with AWS without trying to access any resource)

Reply:
From https://materializeinc.slack.com/archives/CL68GT3AT/p1699995169275279?thread_ts=1699995043.914799&cid=CL68GT3AT, we can use `aws sts get-caller-identity` to validate the credentials. Thanks @rjobanp!